### PR TITLE
MVVM refactor

### PR DESCRIPTION
**Pull request by agustingodoyc**
**STORY**
Refactor to MVVM
**DESCRIPTION**
refactor to MVVM implementation
**EVIDENCE**
image
Sep-09-2022 16-02-12

### CODE REVIEW

File: Ted Talk/Ted Talk/Model/FromWebParser.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
+//
+//  TedTalkService.swift
+//  Ted Talk
+//
+//  Created by Agustin Godoy Cosser on 07/09/2022.
+//
+
+import Foundation
+
+class FromWebParser: ServiceProtocol {
```
Comment (ezequielvelezglobant) [inline Ted Talk/Ted Talk/Model/FromWebParser.swift]: Why you rename this to FromWebParser? this is not a good name. If it's getting info from a web service should be name something related to that.

File: Ted Talk/Ted Talk/ViewModel/TableViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            self.delegate?.loadData()
+            self.data = result
+        }
+    }
+}
+
+// MARK: - Table View
+
+extension TableViewModel {
+    func getDataRows() -> Int {
```
Comment (ezequielvelezglobant) [inline Ted Talk/Ted Talk/ViewModel/TableViewModel.swift]: this name is not correct getDataRows represent that you get all the rows, not the number of rows

File: Ted Talk/Ted Talk/View/TableViewController.swift [hunk 3]
```diff
@@ -49,67 +31,87 @@ class ViewController: UIViewController {
...
     
     func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
         guard let cell = tableView.dequeueReusableCell(withIdentifier: "tedTalkCell",
-                                                       for: indexPath) as? TedTalkCell else {
+                                                       for: indexPath) as? CellView else {
             return UITableViewCell()
         }
         
-        cell.loadData(.init(talk: data[indexPath.row]))
+        cell.loadData(.init(viewModel.getCell(indexPath.row)))
```
Comment (ezequielvelezglobant) [inline Ted Talk/Ted Talk/View/TableViewController.swift]: the ViewModel should be responsible of create and return a CellModel.

File: Ted Talk/Ted Talk/View/TableViewController.swift [hunk 4]
```diff
@@ -49,67 +31,87 @@ class ViewController: UIViewController {
...
+
+// MARK: - Detail View
+
+extension TableViewController {
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        guard let selectedPath = tableView.indexPathForSelectedRow else {
+            return
+        }
+        if segue.identifier == "detail", let destination = segue.destination as? DetailViewController {
+            destination.tedTalk = .init(viewModel.getCell(selectedPath.row))
```
Comment (ezequielvelezglobant) [inline Ted Talk/Ted Talk/View/TableViewController.swift]: the ViewModel should be responsible of create and return a DetailModel.

**Review by ezequielvelezglobant — CHANGES_REQUESTED** [review]

File: Ted Talk/Ted Talk/View/TableViewController.swift (hunk 4 above)
Comment (agustingodoyc) [reply Ted Talk/Ted Talk/View/TableViewController.swift]: [c6611af](https://github.com/agustingodoyc/iOS_Academy_base_project/pull/6/commits/c6611af708dc364d8f6621f9ee20997b02ad7d4c)

File: Ted Talk/Ted Talk/View/TableViewController.swift (hunk 3 above)
Comment (agustingodoyc) [reply Ted Talk/Ted Talk/View/TableViewController.swift]: [c6611af](https://github.com/agustingodoyc/iOS_Academy_base_project/pull/6/commits/c6611af708dc364d8f6621f9ee20997b02ad7d4c)

File: Ted Talk/Ted Talk/ViewModel/TableViewModel.swift (hunk 2 above)
Comment (agustingodoyc) [reply Ted Talk/Ted Talk/ViewModel/TableViewModel.swift]: [c6611af](https://github.com/agustingodoyc/iOS_Academy_base_project/pull/6/commits/c6611af708dc364d8f6621f9ee20997b02ad7d4c)

File: Ted Talk/Ted Talk/Model/FromWebParser.swift (hunk 1 above)
Comment (agustingodoyc) [reply Ted Talk/Ted Talk/Model/FromWebParser.swift]: [c6611af](https://github.com/agustingodoyc/iOS_Academy_base_project/pull/6/commits/c6611af708dc364d8f6621f9ee20997b02ad7d4c)

File: Ted Talk/Ted Talk/ViewModel/TableViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+import Foundation
+
+class TableViewModel {
+    
+    // MARK: - Properties
+    
+    let dataManager: DataManager = DataManager()
+    
+    var delegate: ViewModelDelegate?
```
Comment (ezequielvelezglobant) [inline Ted Talk/Ted Talk/ViewModel/TableViewModel.swift]: you have to add weak to the delegate to avoid retain cycles!

**Review by ezequielvelezglobant — CHANGES_REQUESTED** [review]

File: Ted Talk/Ted Talk/ViewModel/TableViewModel.swift (hunk 5 above)
Comment (agustingodoyc) [reply Ted Talk/Ted Talk/ViewModel/TableViewModel.swift]: [d93d09f](https://github.com/agustingodoyc/iOS_Academy_base_project/pull/6/commits/d93d09ffb4c04db5a18a16d62902995816907d3f)

**Review by JMVarelaGlobant — CHANGES_REQUESTED** [review]
Please fix errors on ParseTest. Every time before pushing your changes to remote, you need to run the tests to ensure your changes don't break anything in your app

**Review by ezequielvelezglobant — APPROVED** [review]

**Review by ezequielvelezglobant — APPROVED** [review]

**Review by JMVarelaGlobant — APPROVED** [review]

**Pull request merged**